pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go

```diff
@@ -0,0 +1,1339 @@
+// StateReplicate relies on receiving responses to those outgoing messages, we
+// should not miss any transitions. If stale messages received in a step can
+// cause transitions out and back without observation, we can add a monotonic
+// counter for each follower inside Raft (this is just local state at the
+// leader), which will be incremented on every state transition and expose
+// that via the RaftInterface.
+type Ready interface {
+	// Entries represents the new entries that are being added to the log.
+	// This may be empty.
+	//
+	// The second byte of AC entries will encode the priority, so we will have
+	// the priority, size, and index position, which is all that we need. Unlike
+	// v1, where we are doing the accounting in
+	// https://github.com/cockroachdb/cockroach/blob/f601b7b439ced71030bfdb0d9ba9cb4925420569/pkg/kv/kvserver/replica_proposal_buf.go#L1057-L1066
+	// which is messy.
+	Entries() []raftpb.Entry
+	// RetransmitMsgApps returns the MsgApps that are being retransmitted, or
+	// being used to ping the follower. These will never be queued by
+	// replication AC, but it may adjust the priority before sending (due to
+	// priority inheritance, when the Message has non-empty Entries).
+	RetransmitMsgApps() []raftpb.Message
+}
+
+// RaftInterface abstracts what the RangeController needs from the raft
+// package, when running at the leader. It also provides one piece of
+// information that is not in the raft package -- the current connected state
+// of the store, that we will use the RaftTransport to answer.
+//
+// NB: group membership and connectivity information is communicated to the
+// RangeController via a separate channel, as in the v1 implementation, since
+// it depends on information inside and outside Raft. The methods in Raft are
+// the ones RangeController will call.
+//
+// The implementation should not need to acquire Replica.mu since we have not
+// said anything about the relative lock ordering of Replica.mu and the
+// internal mutexes in RangeController.
+type RaftInterface interface {
+	// FollowerState returns the current state of a follower. The value of
+	// nextUpperBound is populated iff in StateReplicate. All entries >=
+	// nextUpperBound have not yet had MsgApps constructed.
+	//
+	// When a follower transitions from {StateProbe,StateSnapshot} =>
+	// StateReplicate, or was already in StateReplicate but was disconnected and
+	// has now reconnected, we can now start trying to send MsgApps. We should
+	// notice such transitions both in HandleRaftEvent and SetReplicas. We
+	// should *not* construct MsgApps for a StateReplicate follower that is
+	// disconnected -- there is no timeliness guarantee on how long a follower
+	// will stay in StateReplicate despite it being down, and by sending such a
+	// follower MsgApps that are not being received we are defeating flow
+	// control (since we will have advanced nextUpperBound).
```

Comment:
Do we intend to stay in `StateReplicate` if the follower is disconnected? Today, there is a `ReportUnreachable` call that puts the flow to `StateProbe`. If this method is not called, raft will be sending `MsgApps` until it reaches max-inflight limits and stalls (there won't be more than one probe because a disconnected follower does not reply to heartbeats).

I'm guessing we're moving control over "raft will be sending `MsgApps` until it reaches max-inflight limits and stalls" to AC. So we may choose not to send any `MsgApps`, not call `ReportUnreachable`, and stay in `StateReplicate`. However, when we reconnect, there is a non-zero chance of a reject and needing to probe.

But I guess we won't need any special handling for this disconnected state because we will unmount the "connected stream", and won't be sending any messages as a result.